vllm/model_executor/parallel_utils/pynccl.py

```diff
@@ -0,0 +1,229 @@
+# ===================== pynccl.py ==================================
+# This file is a pure Python wrapper for the NCCL library.
+# The main purpose is to use NCCL combined with CUDA graph.
+# Before writing this script, we tried the following approach:
+# 1. We tried to use `cupy`, it calls NCCL correctly, but `cupy` itself
+#  often gets stuck when initializing the NCCL communicator.
+# 2. We tried to use `torch.distributed`, but `torch.distributed.all_reduce`
+#  contains many other potential cuda APIs, that are not allowed during
+#  capturing the CUDA graph. For further details, please check
+# https://discuss.pytorch.org/t/pytorch-cudagraph-with-nccl-operation-failed/199366
+# ====================================================
+
+# ===================== import region =====================
+import torch
+import ctypes
+import torch.distributed as dist
+from torch.distributed import ReduceOp
+import datetime
+import os
+import glob
+
+# manually load the nccl library
+_path = os.path.dirname(os.path.abspath(__file__))
+so_file = glob.glob(f"{_path}/../../lib/nvidia/nccl/lib/libnccl.so.*")[0]
```

Comment:
Haven't looked into the whole PR. But I feel having this kind of absolute path for libnccl seems very dangerous. Many cloud providers may ship their own nccl package (e.g., aws for efa support). I'm not sure about what's the effect of this.

Reply:
This is our secrect version of `nccl`, living inside `vllm`, we ship it with our wheel. Please check the modification in `setup.py`.

We can remove this after `nccl` fixes the bug, or after pytorch downgrades to `nccl==2.18`.

Reply:
Expanding the path, you can see it is `/path/to/vllm/lib/nvidia/nccl/lib/libnccl.so.2.18.3`. I created a secrect `lib` folder under `vllm` library, and this `.so` file will come with `vllm` installation.